pkg/cluster/streams.go

```diff
@@ -0,0 +1,349 @@
+	appIds := make([]string, 0)
+	for _, stream := range streams {
+		if !util.SliceContains(appIds, stream.ApplicationId) {
+			appIds = append(appIds, stream.ApplicationId)
+		}
+	}
+
+	return appIds
+}
+
+func (c *Cluster) syncPostgresConfig() error {
+	slots := make(map[string]map[string]string)
+	publications := make(map[string]map[string]acidv1.StreamTable)
+	createPublications := make(map[string]string)
+	alterPublications := make(map[string]string)
+
+	defer func() {
+		if err := c.closeDbConn(); err != nil {
+			c.logger.Errorf("could not close database connection: %v", err)
+		}
+	}()
+
+	desiredPatroniConfig := c.Spec.Patroni
+	if len(desiredPatroniConfig.Slots) > 0 {
+		slots = desiredPatroniConfig.Slots
+	}
+
+	// define extra logical slots for Patroni config
+	for _, stream := range c.Spec.Streams {
+		slot := map[string]string{
+			"database": stream.Database,
+			"plugin":   constants.EventStreamSourcePluginType,
+			"type":     "logical",
+		}
+		slotName := getSlotName(stream.Database, stream.ApplicationId)
+		if _, exists := slots[slotName]; !exists {
+			slots[slotName] = slot
+			publications[slotName] = stream.Tables
+		} else {
+			streamTables := publications[slotName]
+			for tableName, table := range stream.Tables {
+				if _, exists := streamTables[tableName]; !exists {
+					streamTables[tableName] = table
+				}
+			}
+			publications[slotName] = streamTables
+		}
+	}
+
+	if len(slots) > 0 {
+		desiredPatroniConfig.Slots = slots
+	} else {
+		return nil
+	}
+
+	// if streams are defined wal_level must be switched to logical
+	desiredPgParameters := map[string]string{"wal_level": "logical"}
+
+	// apply config changes in pods
+	pods, err := c.listPods()
+	if err != nil || len(pods) == 0 {
+		c.logger.Warningf("could not list pods of the statefulset: %v", err)
+	}
+	for i, pod := range pods {
+		podName := util.NameFromMeta(pods[i].ObjectMeta)
+		effectivePatroniConfig, effectivePgParameters, err := c.patroni.GetConfig(&pod)
+		if err != nil {
+			c.logger.Warningf("could not get Postgres config from pod %s: %v", podName, err)
+			continue
+		}
+
+		_, err = c.checkAndSetGlobalPostgreSQLConfiguration(&pod, effectivePatroniConfig, desiredPatroniConfig, effectivePgParameters, desiredPgParameters)
+		if err != nil {
+			c.logger.Warningf("could not set PostgreSQL configuration options for pod %s: %v", podName, err)
+			continue
+		}
+	}
+
+	// next, create publications to each created slot
+	c.logger.Debug("syncing database publications")
+	for publication, tables := range publications {
+		// but first check for existing publications
+		dbName := slots[publication]["database"]
+		if err := c.initDbConnWithName(dbName); err != nil {
+			return fmt.Errorf("could not init database connection")
+		}
+
+		currentPublications, err := c.getPublications()
+		if err != nil {
+			return fmt.Errorf("could not get current publications: %v", err)
+		}
+
+		tableNames := make([]string, len(tables))
+		i := 0
```

Comment:
why is this counter needed ?

Reply:
to fill the tableNames slice a bit quicker. I [read](https://stackoverflow.com/questions/21362950/getting-a-slice-of-keys-from-a-map) that append (which we usually use :smiley: ) is slower.